### PR TITLE
Screenshot path pattern (closes #1602, #1651, #1974, #1975)

### LINKED ISSUE #1602: Screenshot flexibility request
### Are you requesting a feature or reporting a bug?
Feature

### What is the current behavior?
Screenshot naming conventions are very restricted, command line parameter `-S` has no control at all.

### What is the expected behavior?
1)  Screenshots should have flexible naming conventions.  For example, [Junit Attachment Plugin](https://wiki.jenkins.io/display/JENKINS/JUnit+Attachments+Plugin) uses the `classname` xml attribute to associate the artifacts with the package, or uses absolute references
2)  Fixtures and/or tests should have the ability to individually take screenshots on fail.  Maybe a hook that can be defined per test?  If I have a fixture with:

	```js
	test("Test A", async => {});
    test("Test B", async => {});
    test("Test C", async => {});
	```
	Then I should be able to specify which tests I want to take a screenshot on failure.  Example:
    ```js
	test("Test A", async => {});
    test
      .screenShotOnFail(
		//Some kind of path structure here maybe?      
      )
      ("Test B", async => {});
    test("Test C", async => {});
	```

### Specify your

* operating system:  Windows 7
* testcafe version:  0.16.0
* node.js version:  6.10.1

Similar issues:  #1515 and #1595

### PR DESCRIPTION
**Pull request by miherlosev**

### CODE REVIEW

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
FPR @AndreyBelym @VasilyStrelyaev

File: src/cli/argument-parser.js [hunk 1]
```diff
@@ -18,8 +17,8 @@ const REMOTE_ALIAS_RE          = /^remote(?::(\d*))?$/;
 const DEFAULT_TEST_LOOKUP_DIRS = ['test/', 'tests/'];
 const TEST_FILE_GLOB_PATTERN   = `./**/*@(${Compiler.getSupportedTestFileExtensions().join('|')})`;
 
-const DESCRIPTION = dedent(`
-    In the browser list, you can use browser names (e.g. "ie9", "chrome", etc.) as well as paths to executables.
+const DESCRIPTION = `
```
Comment (AndreyBelym) [inline src/cli/argument-parser.js]: Why no `dedent` here?

Comment (miherlosev) [reply src/cli/argument-parser.js]: Yes, need to revert back this change

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
FPR

**Comment by AndreyBelym** [comment]
@VasilyStrelyaev, please check the new option description: https://github.com/DevExpress/testcafe/pull/2562/files#diff-032d695d390a3643d7f0e829fdb8097cR91

File: src/runner/index.js [hunk 2]
```diff
@@ -198,9 +197,10 @@ export default class Runner extends EventEmitter {
         return this;
     }
 
-    screenshots (path, takeOnFails = false) {
+    screenshots (path, takeOnFails = false, pattern) {
```
Comment (AndreyBelym) [inline src/runner/index.js]: Please, add a default value to make clear that `pattern` is optional.

File: src/cli/argument-parser.js [hunk 3]
```diff
@@ -89,6 +88,7 @@ export default class CLIArgumentParser {
             .option('-r, --reporter <name[:outputFile][,...]>', 'specify the reporters and optionally files where reports are saved')
             .option('-s, --screenshots <path>', 'enable screenshot capturing and specify the path to save the screenshots to')
             .option('-S, --screenshots-on-fails', 'take a screenshot whenever a test fails')
+            .option('-p, --screenshot-path-pattern <pattern>', 'use patterns to compose screenshot file names and paths: ${BROWSER}, ${BROWSER_VERSION}, ${OS}, ${OS_VERSION}, ${USERAGENT}, ${DATE}, ${TIME}, ${FIXTURE}, ${TEST}, ${TEST_INDEX}, ${FILE_INDEX}, ${QUARANTINE_ATTEMPT}')
```
[the remark this answers is not in the document]
Comment (miherlosev) [reply src/cli/argument-parser.js]: Agree. Without full description for some placeholders (`${FILE_INDEX}, ${QUARANTINE_ATTEMPT}`) it's a difficult to guess the replaced value.

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
@testcafe-build-bot retest

**Comment by miherlosev** [comment]
FPR @churkin

File: src/screenshots/index.js [hunk 4]
```diff
@@ -1,56 +1,19 @@
...
-        this.screenshotsPath       = path;
-        this.testEntries           = [];
-        this.screenshotBaseDirName = Screenshots._getScreenshotBaseDirName();
-        this.userAgentNames        = [];
-    }
-
-    static _getScreenshotBaseDirName () {
-        var now = Date.now();
-
-        return moment(now).format('YYYY-MM-DD_hh-mm-ss');
```
Comment (bxt) [inline src/screenshots/index.js]: @miherlosev You also fixed another bug here: the hours were not in 24h format, but are now! 🎉🎉🎉 Thank you!

**Pull request merged**